File: ddmlib/src/main/java/com/android/ddmlib/input/EventHub.java
package com.android.ddmlib.input;

import com.android.ddmlib.adb.*;
import com.android.ddmlib.adb.TimeoutException;
import com.android.ddmlib.utils.Log;

import java.io.IOException;
import java.util.concurrent.*;

/**
 * 总线,负责从所有驱动文件中读出原始事件
 */
class EventHub {

    private static final String TAG = "EventHub";
    private static final boolean DEBUG = false;
    private static final int OFFER_TIMEOUT = 10;//10s
    /**
     *
     */
    private InputManager mContext;
    /**
     * 线程池，用于统一驱动管理Observer
     */
    private ExecutorService executorService = Executors.newCachedThreadPool();
    /**
     * 原始事件的队列
     */
    private BlockingQueue<IRawEvent> rawEventPool = new LinkedBlockingQueue<>();

    public EventHub(InputManager inputManager) {
        mContext = inputManager;
        executorService.submit(new InputDeviceObserver(null));
    }

    /**
     * 获取一个原始事件进行消费,若当前没有可以消费的事件,那么该函数将会block
     *
     * @return 返回null表示出错, 否则将会返回待消费的事件
     */
    IRawEvent takeRawEvent() {
        if (executorService.isShutdown()) {
            return null;
        }
        try {
            return rawEventPool.take();
        } catch (InterruptedException e) {
            Log.d(TAG, "takeRawEvent error  will return null");
        }
        return null;
    }


    private void offerRawEvent(IRawEvent rawEvent) {
        try {
            if(DEBUG)Log.d(TAG,"offerRawEvent:"+rawEvent);
            rawEventPool.offer(rawEvent, OFFER_TIMEOUT, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
            Log.e(TAG, "offerRawEvent is timeout!!");
        }
    }

    /**
     * 关闭eventhub,通常在设备断线时调用
     */
    public void quit() {
        executorService.shutdownNow();
        rawEventPool.clear();
        Log.d(TAG, "quit.." + executorService);
    }

    /**
     * 读取底层事件
     */
    class InputDeviceObserver implements Callable<Void> {
        private InputDevice inputDevice;

        public InputDeviceObserver(InputDevice inputDevice) {
            this.inputDevice = inputDevice;
        }

        @Override
        public Void call() {
            try {
                EventCommand command = new EventCommand();
                command.addFlags(EventCommand.FLAG_SHOW_LABEL | EventCommand.FLAG_QUIET | EventCommand.FLAG_SHOW_TIMESTAMP);
                mContext.getRemoteDevice().executeShellCommand(new SingleLineReceiver() {
                    @Override
                    public void processNewLines(String line) {
                        IRawEvent rawEvent = new PlainTextRawEvent(line, null);
                        offerRawEvent(rawEvent);
                    }
                }, IDevice.NO_TIMEOUT,command.toString(), inputDevice == null ? "" : inputDevice.getDevFile());

            } catch (TimeoutException | AdbCommandRejectedException | ShellCommandUnresponsiveException
                    | IOException e) {
                e.printStackTrace();
            }
            Log.d(TAG, "stop :" + inputDevice.getDevFile());
            return null;
        }
    }
}

File: ddmlib/src/main/java/com/android/ddmlib/input/android/EventHub.kt
package com.android.ddmlib.input.android

import com.android.ddmlib.*
import java.io.IOException
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

internal class EventHub(private val mContext: InputManager) {

    private val executorService = Executors.newCachedThreadPool()

    private val rawEvents = LinkedBlockingQueue<RawEvent>()

    private val futureHashMap = HashMap<String, Future<*>>()

    private val mDevices = HashMap<String, InputDevice>()

    init {
        scanDevice(true)
    }

    fun scanDevice(force: Boolean) {
        if (force || mDevices.isEmpty()) {
            try {
                mContext.androidDevice.executeShellCommand(object : OneLineReceiver() {
                    internal var sb = ArrayList<String>()
                    override fun processNewLines(line: String) {
                        if (DEBUG) Log.d(javaClass.name, line)
                        if (line.startsWith("add device")) {
                            if (sb.size > 0) {
                                putDevice(sb)
                                sb.clear()
                            }
                        }
                        sb.add(line.trim { it <= ' ' })
                        //                    if(DEBUG)Log.d(getClass().getName(), sb.toString());
                    }

                    override fun done() {
                        putDevice(sb)
                        sb.clear()
                    }
                }, TIMEOUT_SEC, TimeUnit.SECONDS, Command.GETEVENT_GETDEVICE)
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: AdbCommandRejectedException) {
                e.printStackTrace()
            } catch (e: ShellCommandUnresponsiveException) {
                //            e.printStackTrace();
            } catch (e: TimeoutException) {
                //            e.printStackTrace();
            } finally {

            }
        }
        //        result.add(new InputDevice((List<String>) sb.clone(), dev));
        //        return new ArrayList<>(devList.values());
    }

    fun putDevice(sb: ArrayList<String>) {
        val tempDev = InputDevice(sb.clone() as List<String>)
        if (!mDevices.containsKey(tempDev.devFile)) {
            mDevices.put(tempDev.devFile!!, tempDev)
            addDevice(tempDev)
        }
    }

    fun addDevice(tempDev: InputDevice) {
        if (futureHashMap.containsKey(tempDev.devFile)) {
            if (futureHashMap[tempDev.devFile]?.isDone()!!) {

            } else {
                return
            }
        }

        val submitFuture = executorService.submit<Void> {
            try {
                mContext.androidDevice.executeShellCommand(object : OneLineReceiver() {
                    override fun processNewLines(line: String) {
                        val rawEvent = RawEvent(line, tempDev.devFile)
                        rawEvents.add(rawEvent)
                    }
                }, -1, Command.GETEVENT_WHATCH_TEXT_EVENT, tempDev.devFile!!)

            } catch (e: TimeoutException) {
                e.printStackTrace()
            } catch (e: AdbCommandRejectedException) {
                e.printStackTrace()
            } catch (e: ShellCommandUnresponsiveException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }

            null
        }
        futureHashMap.put(tempDev.devFile!!, submitFuture)
    }


    fun removeMonitorDevice(): Boolean {
        return false
    }

    val event: RawEvent
        @Throws(InterruptedException::class)
        get() = rawEvents.take()


    val devices: ArrayList<InputDevice>
        get() = ArrayList(mDevices.values)

    companion object {


        //    private LinkedHashMap<String>

        //    private IDevice mAndroidDevice;

        private val TAG = "EventHub"
        private val DEBUG = false

        private val TIMEOUT_SEC = 2L
    }
}
